### ISSUE TITLE
Element upgrade ordering

### ISSUE THREAD

**Issue by justinfagnani** [WICG/webcomponents#923]
The scoped custom elements proposal needs to define the upgrade order of custom elements within shadow roots that have a scoped registry.

We've discussed the question of scoped registry upgrade ordering in other issues, but it deserves its own issue so we can more easily track the discussion and consensus.

The discussion on ordering starts here: https://github.com/WICG/webcomponents/issues/716#issuecomment-677071243 and is also brought up here: https://github.com/WICG/webcomponents/pull/865#issuecomment-699089641 with the second link having some more detailed discussion.

To reiterate, the two main options seem to be:
1. Keep document order upgrading.
2. Order upgrades by the order that shadow roots were associated with the registry, then tree order within those roots.

It _seemed_ from the discussion in the last face-to-face and in #865 that if disconnected trees are not upgraded and registries are restricted to one document that keeping a list of shadow roots associated with a registry - in the order they were created, and upgrading only those roots was preferable. I did want open this to double check with @rniwa @annevk and @emilio especially.

**Comment by annevk** [comment]
Yeah, the argument in favor of 2 is performance. I don't think I explained the argument of 1 during the call, but it's consistency with the global registry.

Breaking consistency for the sake of performance is acceptable, but it also comes with additional bookkeeping as you note. (I'm wondering if https://github.com/WICG/webcomponents/issues/907#issuecomment-825266773 applies here as well, but I'm not familiar enough with the specifics.)

**Comment by rniwa** [comment]
Note that if we went with (2), we still can't allow upgrading of elements in a disconnected subtree for the same reasons as I had stated in https://github.com/WICG/webcomponents/issues/419

**Comment by justinfagnani** [comment]
@rniwa agreed. I think we mentioned this in the f2f as well as not wanting to diverge from global registry behavior.

**Comment by trusktr** [comment]
Element upgrade order with a document registry already changes depending on various factors (f.e. whether elements come before a script tag that runs all customElements.define calls, etc), so authors already _must_ design their elements such that upgrade order _does not matter_ if they wish for their elements to be robust and to work properly regardless in which order some end user may load things.

Threfore, specifying a specific order for ShadowRoot registries that does not match document registries is totally acceptable, and an author's robust elements will probably already just work regardless. (It is a huge headache to figure how to write that robustness though).

**Comment by keithamus** [comment]
[WCCG had their spring F2F](https://github.com/WICG/webcomponents/issues/978#issuecomment-1516897276) in which this was discussed. I'd like to point out that you can read the full notes of the discussion (https://github.com/WICG/webcomponents/issues/978#issuecomment-1516897276) in which this was discussed, heading entitled "Scoped Custom Element Registries".

**Comment by rniwa** [comment]
2023 TPAC F2F resolution: We would use the ordered set of weak references to documents to which a shadow root was created with a given custom element registry. When a shadow root is adopted from one document to another, that document is inserted into this ordered set. We would ignore any documents without browsing context or any disconnected trees at the time of upgrade call. In each document, all shadow trees are visited in the document order.

**Comment by xiaochengh** [comment]
I'm [implementing](https://chromium-review.googlesource.com/c/chromium/src/+/4888750) scoped upgrade in Chromium, and found that the upgrade order can be made even simpler:
- Elements in different documents are upgraded in frame tree order, namely, elements in a frame attached earlier will be upgraded earlier
- Everything else remains the same: upgrade in tree order in the same document, don't upgrade disconnected elements, and don't upgrade elements without a frame

And that doesn't require complicated index structure or expensive full document traversals. We only need to maintain the set of candidate elements for each scoped registry, and when upgrading, perform a bottom-up traversal from those candidates to the root to build a minimal subtree containing every one of them, so that we can sort them.

In fact, the current (non-scoped) registry in Chromium is already doing that, and it's straightforward to extend that to work for multiple tree scopes and multiple documents.

Since there's no implementation issues, I think we should simplify the resolution into using frame tree order when there are multiple documents.

**Comment by annevk** [comment]
The concern we discussed at the meeting was that going through all documents belonging to the same agent would be slow. I'm not even sure we have a well-defined order for all documents belonging to the same agent. (You can have multiple frame trees after all.)

**Comment by smaug----** [comment]
What is "frame tree order"? Is that iframes and such? There is no order between iframes especially if they are in different documents.

**Comment by rniwa** [comment]
Yeah, there could be multiple frame trees (i.e. multiple top-level documents) for a give agent so I don't see how it would be possible to totally order documents based on the frame free ordering.

**Comment by xiaochengh** [comment]
The "frame tree" I'm thinking about is: since each `CustomElementRegistry` is created in a global object `Window`, we can (conceptually) traverse from this `Window` through the [`parent`](https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-parent) and [array index](https://html.spec.whatwg.org/multipage/nav-history-apis.html#windowproxy-getownproperty) getters and get a tree of windows. And the tree order of this tree is [well-defined as the insertion order](https://html.spec.whatwg.org/multipage/nav-history-apis.html#windowproxy-getownproperty). 

I assume there's no way to use a registry in an active document outside this tree, is that correct? If so then we can use this tree.

**Comment by smaug----** [comment]
But the point is that you can use the CustomElementRegistry in another top level window, or in any iframe there.

**Comment by xiaochengh** [comment]
Maybe I missed something obvious, but could you give an example?

Also, if that's possible, I assume these top level windows must be in a bigger tree (since that's how it's implemented in Chromium), whose tree order can still be used here.

**Comment by annevk** [comment]
If you use `window.open()` or `<a target>`. There is some kind of order there too, for sure, but we don't have any existing infrastructure around looping through it all. It's also not immediately clear to me we want to make that order observable.

**Comment by xiaochengh** [comment]
Thanks! I'll add that to the test cases.

Agreed that we probably don't want to expose the tree order implied by `window.opener`.

How about using document creation time order? This is still easy to implement without performance issues, easy to explain to developers and does not require any additional index structure.

**Comment by annevk** [comment]
Document creation-time is also not something we expose today or use as a primitive as far as I know. What's wrong with the set approach we discussed at the meeting?

**Comment by xiaochengh** [comment]
1. I found it harder to implement in Chromium. It requires adding new index structures, while what I proposed can fully make use of existing structures in Chromium.
And document creation time order shouldn't be hard to implement in other engines (if I read the code correctly):
   - WebKit currently doesn't maintain any data structure for upgrades but  [traverses the full document](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/dom/CustomElementRegistry.cpp#L96), so it's free to add any index structure
   - [Gecko's current implementation](https://github.com/mozilla/gecko-dev/blob/23808d46cde6155213b1230675b00a0a426f466e/dom/base/CustomElementRegistry.cpp#L712) seems similar to Chromium, so it doesn't require any additional index structure for document creation time order, either.

2. I think an ordering is simpler for spec and developers if it depends less on the past states

Since what we want is just a fixed ordering without implementation or performance issues, if there are multiple orderings to choose, we should choose the simplest one.

> Document creation-time is also not something we expose today or use as a primitive as far as I know

I was inspired by [this](https://html.spec.whatwg.org/multipage/nav-history-apis.html#windowproxy-getownproperty), which uses child frame insertion time but doesn't quite expose it as a primitive:

> [Sort](https://infra.spec.whatwg.org/#list-sort-in-ascending-order) children in ascending order, with navigableA being less than navigableB if navigableA's [container](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-container) was inserted into W's [associated Document](https://html.spec.whatwg.org/multipage/nav-history-apis.html#concept-document-window) earlier than navigableB's [container](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-container) was.

Since we can use frame insertion time, maybe we can also use document creation time in a similar manner?

**Comment by annevk** [comment]
> I was inspired by [this](https://html.spec.whatwg.org/multipage/nav-history-apis.html#windowproxy-getownproperty)

I see, I consider that extremely fragile and not well tested. A lot of the words there are not defined in terms of well-understood primitives.

**Comment by rniwa** [comment]
FWIW, traversing documents in their creation order isn't a thing so we'd need a way to sort documents in that order, which could be quite expensive. Furthermore, each scoped registry needs to remember the set of documents that contain a shadow tree which uses the registry anyway since otherwise UA would have to traverse through all documents in the creation order and filter those that do (e.g. by having a set of all scoped registries per document).

My preference is to still use the ordering we agreed to during TPAC since it's a lot simpler to spec and implement.

**Comment by smaug----** [comment]
I agree. The approach we discussed during TPAC is indeed simple, to spec and implement and to _understand_.

**Comment by xiaochengh** [comment]
All right, I've [implemented](https://chromium.googlesource.com/chromium/src/+/ce0908e2f55bdd620e46fb68adcb0065d44cb06f) it in Chromium following the TPAC resolution's upgrade ordering.

It does need some new structures, but much less than I thought. So everything is fine, and I suppose we just need some explainer/spec edits before closing this issue?

**Comment by annevk** [comment]
For clarity, https://github.com/whatwg/html/issues/10854 addresses this.
